### PR TITLE
Replace exponential with torch.log in generation sampling

### PR DESCRIPTION
**Pull request by felipemello1**
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [x] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

[Starting in torch 2.7.0.dev20250202, our tests began to fail.](https://github.com/pytorch/torchtune/pull/2362). After investigating, we found that [q = torch.empty_like(probs).exponential_(1) ](https://github.com/pytorch/torchtune/blob/3a2179cd22049ad873a81d6d5a1e7e3b5f8b8c80/torchtune/generation/_generation.py#L62) was the culprit.

```
import torch

torch.manual_seed(42)
print("torch.rand(5):", torch.rand(5))
print("torch.empty(5).exponential_(1):", torch.empty(5).exponential_(1))
```

In nightlies 2.8.0.dev20250401+cu126:
NOTE (pinning to 2.7.0.dev20250201 still works, so [dev20250202](https://github.com/pytorch/pytorch/commit/738ebb45cad7cef7fa0878ce4f835be59d186a94#diff-ecbd0bdf6ee8fc4eab27af1ec0b4369962dd340547451ac2f396b8f12f2f41bc) seems to be the cause):
```
torch.rand(5): tensor([0.8823, 0.9150, 0.3829, 0.9593, 0.3904])
torch.empty(5).exponential_(1): tensor([0.5809, 0.0436, 2.0548, 1.9680, 0.3315])
```

In stable 2.6.0:
```
torch.rand(5): tensor([0.8823, 0.9150, 0.3829, 0.9593, 0.3904])
torch.empty(5).exponential_(1): tensor([1.6459, 0.4294, 0.0677, 1.3809, 0.7803])
```

### Fix

To mitigate that, we can replace the function with rand_like+log to sample from an:
```
probs = torch.ones(5)
u = torch.rand_like(probs)  # Uniform random samples
q = -torch.log(1 - u)  # Transform to exponential distribution
```

this is [what exponential does](https://github.com/pytorch/pytorch/blob/e62d958f02afd74e4d4c023fa97d6c7966e00e26/torch/_refs/__init__.py#L6132), but for some reason random seed it not working on it anymore on nightlies.

### Test

1) Update stop tokens
2) Updated random gibberish generated
3) Remove FSDPTest, which was breaking in nigthlies

### CODE REVIEW

**Review by joecummings — APPROVED** [review]
kewl
